### ISSUE TITLE
twitter publish crash: unhashable type: 'dict'

### ISSUE THREAD

**Issue by snarfed** [snarfed/bridgy#419]
discovered by @diplix (https://www.brid.gy/twitter/diplix). [log](https://www.brid.gy/log?start_time=1435068803&key=aglzfmJyaWQtZ3lyVgsSDVB1Ymxpc2hlZFBhZ2UiL2h0dHA6Ly93aXJyZXMubmV0L2FydGljbGUvYXJ0aWNsZXZpZXcvNzkyMy8xLzYvDAsSB1B1Ymxpc2gYgICAgIDkkQoM):

```
I Params: [(u'source', u'http://wirres.net/7923'), (u'target', u'http://brid.gy/publish/twitter'), (u'bridgy_omit_link', u'')]
D Resolved http://wirres.net/7923 to http://wirres.net/article/articleview/7923/1/6/
I Source: https://www.brid.gy/twitter/diplix , features [u'publish', u'listen'], status enabled
D Publish entity: aglzfmJyaWQtZ3lyVgsSDVB1Ymxpc2hlZFBhZ2UiL2h0dHA6Ly93aXJyZXMubmV0L2FydGljbGUvYXJ0aWNsZXZpZXcvNzkyMy8xLzYvDAsSB1B1Ymxpc2gYgICAgIDkkQoM
I Starting new HTTP connection (1): wirres.net
D "GET /article/articleview/7923/1/6/ HTTP/1.1" 200 None
D Parsed microformats2: {
...
"items": [
{
"type": [
"h-entry"
],
"properties": {
"category": [
""
],
"updated": [
"2015-06-23T16:13:01+0200"
],
"name": [
"",
""
],
"author": [
{
"type": [
"h-card"
],
"properties": {
"url": [
"http://wirres.net"
],
"photo": [
"http://root.wirres.net/03022687630_h2.jpg"
],
"name": [
"felix schwenzel"
]
},
"value": " felix schwenzel"
}
],
"url": [
"http://wirres.net/article/articleview/7923/1/6/"
],
"content": [
{
"html": "\n<div class=\"h-cite p-repost-of repost-of linkitemq key\" id=\"link-b2caaf8f031ec16506d51e1e4bed061a\">\n<p><img http:=\"\" src=\"&lt;/p&gt;\n&lt;link href=\" target=\"_blank\" text=\"http://root.wirres.net/favicon.php?url=https%3A%2F%2Fwww.flickr.com%2Fphotos%2F134213681%40N05%2F19035489016%2Fin%2Fphotostream%2F\"/>\" align=\"top\" width=\"16\" height=\"16\" alt=\"*\" /&gt;\u00a0<a href=\"\n&lt;link href=\" https:=\"\" target=\"_blank\" text=\"https://www.flickr.com/photos/134213681@N05/19035489016/in/photostream/\"></a>\" class=\"linktitle u-url space p-name\"&gt;IMG_6270_1</p>\n<p><a class=\"h-card vcard author p-author u-url p-name fn value name\" href=\"&lt;/p&gt;\n&lt;link href=\" https:=\"\" target=\"_blank\" text=\"https://www.flickr.com/photos/134213681@N05/\"></a>\"&gt;Nebenan. Freundliche Internet-Konferenz:</p>\n<blockquote class=\"e-content\"><p>Foto: Amelie Nerger, @nergera CC-Lizenz BY-NC-SA</p></blockquote>\n</div>\n<p><a href=\"https://www.brid.gy/publish/facebook\"></a><a href=\"https://www.brid.gy/publish/twitter\"></a></p>\n<a name=\"mehr\"></a>\n",
"value": "\n\n\" align=\"top\" width=\"16\" height=\"16\" alt=\"*\" />\u00a0\" class=\"linktitle u-url space p-name\">IMG_6270_1\n\">Nebenan. Freundliche Internet-Konferenz:\nFoto: Amelie Nerger, @nergera CC-Lizenz BY-NC-SA\n\n\n\n"
}
],
"published": [
"2015-06-23T16:13:01+0200"
],
"repost-of": [
"https://www.flickr.com/photos/134213681@N05/19035489016/in/photostream/",
{
"type": [
"h-cite"
],
"properties": {
"url": [
{
"type": [
"h-card"
],
"properties": {
"url": [
"http://wirres.net/article/articleview/7923/1/6/</p>;\n<link href="
],
"name": [
""
]
},
"value": ""
}
],
"content": [
{
"html": "<p>Foto: Amelie Nerger, @nergera CC-Lizenz BY-NC-SA</p>",
"value": "Foto: Amelie Nerger, @nergera CC-Lizenz BY-NC-SA"
}
],
"name": [
{
"type": [
"h-card"
],
"properties": {
"url": [
"http://wirres.net/article/articleview/7923/1/6/</p>;\n<link href="
],
"name": [
""
]
},
"value": ""
}
],
"author": [
{
"type": [
"h-card"
],
"properties": {
"url": [
"http://wirres.net/article/articleview/7923/1/6/</p>;\n<link href="
],
"name": [
""
]
},
"value": ""
}
]
},
"value": "\n\" align=\"top\" width=\"16\" height=\"16\" alt=\"*\" />\u00a0\" class=\"linktitle u-url space p-name\">IMG_6270_1\n\">Nebenan. Freundliche Internet-Konferenz:\nFoto: Amelie Nerger, @nergera CC-Lizenz BY-NC-SA\n"
}
]
}
},
{
"type": [
"h-card"
],
"properties": {
"name": [
"felix schwenzel"
],
"author": [
"felix schwenzel"
],
"url": [
"http://wirres.net/kontakt/"
],
"photo": [
"http://root.wirres.net/03022687630_h2.jpg"
],
"note": [
"wirres.net ist ein publizistisches angebot von felix schwenzel."
],
"logo": [
"http://wirres.net/images/dogpoo-original-114-transparent.png"
],
"nickname": [
"ix"
]
}
}
],
...
]
}
W Error: unhashable type: 'dict'
Traceback (most recent call last):
  File "/base/data/home/apps/s~brid-gy/3.384978554043730343/publish.py", line 195, in _run
result = self.attempt_single_item(item)
  File "/base/data/home/apps/s~brid-gy/3.384978554043730343/publish.py", line 247, in attempt_single_item
obj = microformats2.json_to_object(item)
  File "/base/data/home/apps/s~brid-gy/3.384978554043730343/activitystreams/microformats2.py", line 259, in json_to_object
for field in ('like', 'like-of', 'repost', 'repost-of', 'in-reply-to')))
  TypeError: unhashable type: 'dict'
```

**Comment by snarfed** [comment]
looks like the problem is that `items[0].properties.url` is a compound object, which we handle, but its `properties.url` is _another_ compound object, which we don't handle well.

the fix is probably just to make [`microformats2.get_string_urls()`](https://github.com/snarfed/activitystreams-unofficial/blob/master/activitystreams_unofficial/microformats2.py#L44) recursive. :P

**Comment by diplix** [comment]
i don’t understand a word. but if there’s something i can do on my side to fix this: i’m happy to help.
(like a more compliant h-entry-structure?)

**Comment by snarfed** [comment]
thanks for the offer! bridgy should never crash though, regardless of your page. we'll fix it.

**Comment by kylewm** [comment]
Agree Bridgy shouldn't give a 500, but I think this is a case where it should give a helpful error message rather than trying to cope with any possible nesting of inputs

@diplix I suspect this link should not have `h-card vcard author p-author`, since it is a permalink to the original post and not an h-card describing the author.

```html
<a class="h-card vcard author p-author u-url p-name fn value name" href="https://www.flickr.com/photos/134213681@N05/">Nebenan. Freundliche Internet-Konferenz</a>
```

**Comment by diplix** [comment]
nope, i tend to mees stuff like this up, but this one is actually a profile link: https://www.flickr.com/photos/134213681@N05
the permalink in this article is this: https://www.flickr.com/photos/134213681@N05/19035489016/in/photostream/
   <a href="https://www.flickr.com/photos/134213681@N05/19035489016/in/photostream/" class="linktitle u-url space p-name">IMG_6270_1</a><

**Comment by kylewm** [comment]
Oh, my mistake. Then you should remove `u-url` from the author link, so that it doesn't look like the h-entry permalink to parsers.

**Comment by diplix** [comment]
your probably right.
i copied the markup from aaron parecki (i.e. https://aaronparecki.com/notes/2015/06/13/1/) who represents h-cards just like that. however, he doesn’t nest the h-card in the e-content scope like i do, which actually seems to confuse parsers as you said. so i’ll go without the u-url in future posts and hope that helps.

**Comment by kylewm** [comment]
Aha, there's a subtle but important difference in aaronpk's markup. The properties he uses are nested:

- h-entry
  - p-repost-of h-cite
    - p-author h-card
      - u-url
      - p-name

so u-url and p-name are inside the h-card and are interpreted as properties *of* the h-card

whereas your post has (effectively)

- h-entry
  - p-repost-of h-cite
    - p-author h-card
    - u-url: https://www.flickr.com/photos/134213681@N05/
    - p-name: Nebenan. Freundliche Internet-Konferenz

because u-url and p-name are not inside the p-author h-card, they associate with their outer p-repost-of h-cite.

And it's totally fine to have any and all of this inside the e-content. I actually like doing that way, so parsers that don't know about reposts, and only display the content will still display something sensical.

**Comment by snarfed** [comment]
fixed by snarfed/granary@3344a20236a24099508b2eb620d232505dfe4b9a

(...and oh man "granary" is so much nicer to type than "activitystreams-unofficial" :P)
